### PR TITLE
Add optee os arch and product

### PR DESCRIPTION
**Pull request by vingu-linaro**
This P-R integrates the changes that we have made in SCP firmware in order to make it run as a pseudo Trusted Application in OP-TEE env.

SCP-Firmware is integrated as a sub-module of op_tee-os. this PR has been tested with : https://github.com/vingu-linaro/optee_os/tree/scmi-over-optee.

This has been tested on FVP arm64 and stm32.

### CODE REVIEW

**Comment by vingu-linaro** [comment]
We are still using makefile instead of cmake but wanted to shared the patchset to get 1st feedback while working on integration with cmake

**Comment by tarek-arm** [comment]
Can we move the optee-specific modules from common modules to the optee product modules?

**Comment by vingu-linaro** [comment]
> Would it be possible to provide a test environment which we can add to the CI to make sure this is not going to regress?

Do you mean having a FVP or Qemu based setup to boot and run some tests ?

**Comment by tarek-arm** [comment]
Jenkins: retest this please

File: module/optee_mhu/include/mod_optee_mhu.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#define OPTEE_MHU_SHMEM_SIZE 128
+
+/*!
+ * \brief Channel type
+ *
+ * \details Defines the role of an entity in a channel
+ */
+enum mod_optee_mhu_channel_type {
+    /*! Master channel */
+    MOD_OPTEE_MHU_CHANNEL_TYPE_MASTER,
```
Comment (tgonzalezorlandoarm) [inline module/optee_mhu/include/mod_optee_mhu.h:24]: This is a deprecated term as it doesn't conform to our inclusive language policy

File: module/optee_mhu/include/mod_optee_mhu.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    /*! Slave channel */
+    MOD_OPTEE_MHU_CHANNEL_TYPE_SLAVE,
```
Comment (tgonzalezorlandoarm) [inline module/optee_mhu/include/mod_optee_mhu.h:27]: This is a deprecated term as it doesn't conform to our inclusive language policy

File: CMakeLists.txt [hunk 3]
```diff
@@ -141,7 +141,7 @@ check_ipo_supported(RESULT SCP_IPO_SUPPORTED)
 cmake_dependent_option(
     SCP_ENABLE_IPO
     "Enable the interprocedural optimization (IPO) if supported?"
-    "${SCP_ENABLE_IPO_INIT}" "DEFINED SCP_ENABLE_IPO_INIT" "${SCP_ENABLE_IPO}")
+    "${SCP_ENABLE_IPO_INIT}" "DEFINED SCP_ENABLE_IPO_INIT" TRUE)
```
Comment (tgonzalezorlandoarm) [inline CMakeLists.txt:144]: We don't intend to enable IPO by default, hence, it's being enabled/disabled by setting SCP_ENABLE_IPO_INIT in product specific Firmware.cmake file. From what I understand, this change causes IPO to be enabled by default when SCP_ENABLE_IPO_INIT is not defined. 
Some platforms do not behave as intended when IPO is enabled, so this shouldn't be the case.

**Review by tgonzalezorlandoarm — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (vingu-linaro) [reply CMakeLists.txt:144]: yes, That's a leftover while testing the change below which should have been removed

File: module/optee_mhu/include/mod_optee_mhu.h (hunk 1 above)
Comment (tgonzalezorlandoarm) [reply module/optee_mhu/include/mod_optee_mhu.h:24]: yes, requester/completer is the preferred renaming :)

**Comment by vingu-linaro** [comment]
Hi,

Do you have comment for the lats version that I pushed ?

**Comment by vingu-linaro** [comment]
Hi Al,

I have pushed a new version of the PR to add the support of optee PTA scmi server.
Main changes:
-  created a optee dir with optee related module and a .md file that describe the architecture of  SCP-firmware running in a OPTEE PTA
- created a mock_ppu module to replace the vppu in optee-fvp product.
- normally fixed all typo, spacing and tabs issues

I have not yet make public the associated test plan as i still need to finish the the roadmap. The goal is to include a CI loop with scmi conformance test and some unitary tests.

**Comment by vingu-linaro** [comment]
Hi All,

I have rebased on latest commit to fix merge conflict du to module list reordering. I have also cleanup the fvp optee product

**Review by tarek-arm — CHANGES_REQUESTED** [review]
It is not clear to me how to build the SCP firmware as a library!
Is this a new target which we need to list here:
https://github.com/ARM-software/SCP-firmware/blob/master/Makefile.cmake#L197
Can you please add some documentation to explain?

**Review by tarek-arm — APPROVED** [review]

File: CMakeLists.txt [hunk 4]
```diff
@@ -151,7 +151,7 @@ endif()
 
 cmake_dependent_option(
     CMAKE_INTERPROCEDURAL_OPTIMIZATION "Enable interprocedural optimization?"
-    TRUE "SCP_IPO_SUPPORTED" FALSE)
+    TRUE "SCP_IPO_SUPPORTED;SCP_ENABLE_IPO" FALSE)
```
Comment (tarek-arm) [inline CMakeLists.txt:154]: This check was done in lines 146 to line 149. Why do we need to do it again here?

**Review by tarek-arm — CHANGES_REQUESTED** [review]

**Review by tarek-arm — CHANGES_REQUESTED** [review]

File: product/optee-stm32mp1/fw/config_scmi_clocks.c [hunk 5]
```diff
@@ -0,0 +1,254 @@
+/*
```
Comment (tarek-arm) [inline product/optee-stm32mp1/fw/config_scmi_clocks.c:1]: Please change SMCI to SCMI in the title  "product/optee-stm32mp1: expose SMCI clocks"

**Review by tarek-arm — CHANGES_REQUESTED** [review]

File: module/msg_smt/src/mod_msg_smt.c [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+
+static struct smt_ctx smt_ctx;
+
+/*
+ * SCMI Transport API
+ */
+static int smt_get_secure(fwk_id_t channel_id, bool *secure)
+{
+    if (secure == NULL) {
+        assert(false);
```
Comment (leandro-arm) [inline module/msg_smt/src/mod_msg_smt.c:80]: Please use the provided assert functions from framework `fwk_assert`

File: module/msg_smt/src/mod_msg_smt.c [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+
+    return FWK_SUCCESS;
+}
+
+static int smt_get_max_payload_size(fwk_id_t channel_id, size_t *size)
+{
+    struct smt_channel_ctx *channel_ctx;
+
+    if (size == NULL) {
+        assert(false);
```
Comment (leandro-arm) [inline module/msg_smt/src/mod_msg_smt.c:95]: ditto for all occurrences

File: module/msg_smt/src/mod_msg_smt.c [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+    if ((payload == NULL)                         ||
+        (offset  > channel_ctx->max_payload_size) ||
+        (size > channel_ctx->max_payload_size)    ||
+        ((offset + size) > channel_ctx->max_payload_size)) {
+
+        assert(false);
+        return FWK_E_PARAM;
+    }
+
+    if (!channel_ctx->locked)
```
Comment (leandro-arm) [inline module/msg_smt/src/mod_msg_smt.c:173]: all single-line if conditions should be rounded by braces {}

File: module/optee/mbx/include/mod_optee_mbx.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    /*! Identifier of the driver API to bind to */
+    fwk_id_t driver_api_id;
+};
+
+/*!
+ * \brief Interface to exchange message with OP-TEE
+ */
+
+/* Return number of MBX device registered in the paltform */
```
Comment (leandro-arm) [inline module/optee/mbx/include/mod_optee_mbx.h:29]: This should be exposed through an API structure

File: module/optee/console/src/mod_optee_console.c [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    char *restrict ch)
+{
+    *ch =  'A';
+
+    return FWK_SUCCESS;
+}
+
+static int mod_console_io_putch(const struct fwk_io_stream *stream, char ch)
+{
+    console_putc(ch);
```
Comment (leandro-arm) [inline module/optee/console/src/mod_optee_console.c:57]: where is this created?
I don't know if it is helpful but `stdio` module integrates `fprintf` functions from stdio library

File: module/optee/reset/src/mod_optee_reset.c [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        return FWK_E_ACCESS;
+    }
+
+    /* Whatever the reset_state set, we consider a unique context loss mode */
+    if (reset_state)
+        DMSG("Override requested SCMI reset state %#"PRIx32, reset_state);
+
+    switch (mode) {
+    case MOD_RESET_DOMAIN_MODE_EXPLICIT_ASSERT:
+        FMSG("SCMI reset %u: assert(%s)",
```
Comment (leandro-arm) [inline module/optee/reset/src/mod_optee_reset.c:70]: Ditto for all occurrences

File: product/optee-fvp/fw/Firmware.cmake [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+set(SCP_FIRMWARE_TARGET "scmi-fw")
+
+set(SCP_ARCHITECTURE "optee")
+
+set(SCP_TOOLCHAIN_INIT "GNU")
+
+set(CMAKE_BUILD_TYPE "Release")
+
+#set(SCP_GENERATE_FLAT_BINARY_INIT FALSE)
```
Comment (leandro-arm) [inline product/optee-fvp/fw/Firmware.cmake:22]: Please remove commented code

**Review by leandro-arm — CHANGES_REQUESTED** [review]
Please check my suggestions

File: CMakeLists.txt (hunk 4 above)
Comment (vingu-linaro) [reply CMakeLists.txt:154]: line 139 check_ipo_supported(RESULT SCP_IPO_SUPPORTED) can have already set SCP_IPO_SUPPORTED.
but the product doesn't want to use it

File: CMakeLists.txt [hunk 13]
```diff
@@ -561,6 +567,18 @@ if(SCP_FIRMWARE)
...
+set(SCP_FIRMWARE_LIB "${SCP_FIRMWARE_TARGET}-all")
+
+# merged libraries is only built when `make ${SCP_FIRMWARE_LIB}` is issued
+add_custom_target(${SCP_FIRMWARE_LIB}
+    # COMMAND/COMMENT must be upper case
+        COMMAND ${CMAKE_AR} rcT $<TARGET_FILE_DIR:${SCP_FIRMWARE_TARGET}>/lib${SCP_FIRMWARE_LIB}.a ${scp_libraries}
+        COMMENT " Merging all libraries in a single lib${SCP_FIRMWARE_LIB}.a"
+)
+
+add_dependencies(${SCP_FIRMWARE_LIB} ${SCP_FIRMWARE_TARGET})
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply CMakeLists.txt:584]: what we are doing for optee_os:

cmake -S $(scmi-server-scp-path) -B $(scmi-server-out-path)/build $(scmi-server-flags-y)
make -C $(scmi-server-out-path)/build scmi-fw-all

Our firmware target is named set(SCP_FIRMWARE_TARGET "scmi-fw")
and it's not an executable but a library and we use the library libscmi-fw-all.a in optee-os

we are not using the make wrapper https://github.com/ARM-software/SCP-firmware/blob/master/Makefile.cmake#L197

Where is the best place to put the explanation  ? https://github.com/ARM-software/SCP-firmware/blob/master/doc/cmake_readme.md

**Comment by vingu-linaro** [comment]
> Can you please add documentation under /arc/optee/doc? Explain which systems this configuration should work for. Is it possible also to run the server in both OP-TEE as well as in a micro-controller? We need to explain how the system is supposed to work. A block diagram will also be helpful. You can show the interactions between OSPM --> OP-TEE --> SCMI server --> hardware

I have added a design overview here: https://github.com/vingu-linaro/SCP-firmware/blob/add-optee-os-arch/module/optee/doc/module_optee_design.md. Should I move it tin arch/none/optee instead ?

File: product/optee-stm32mp1/fw/config_scmi_clocks.c (hunk 5 above)
Comment (vingu-linaro) [reply product/optee-stm32mp1/fw/config_scmi_clocks.c:1]: ok

File: module/msg_smt/src/mod_msg_smt.c (hunk 6 above)
Comment (vingu-linaro) [reply module/msg_smt/src/mod_msg_smt.c:80]: ok

File: module/msg_smt/src/mod_msg_smt.c (hunk 8 above)
Comment (vingu-linaro) [reply module/msg_smt/src/mod_msg_smt.c:173]: yes, I missed this one

File: module/optee/console/src/mod_optee_console.c (hunk 10 above)
Comment (vingu-linaro) [reply module/optee/console/src/mod_optee_console.c:57]: console_putc() is the optee-os function to output on the optee console

File: module/optee/reset/src/mod_optee_reset.c [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+    ctx = &module_ctx.dev_ctx_table[fwk_id_get_element_idx(dev_id)];
+
+    if (!is_exposed(ctx)) {
+        return FWK_E_ACCESS;
+    }
+
+    /* Whatever the reset_state set, we consider a unique context loss mode */
+    if (reset_state)
+        DMSG("Override requested SCMI reset state %#"PRIx32, reset_state);
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply module/optee/reset/src/mod_optee_reset.c:66]: yes

File: product/optee-fvp/fw/Firmware.cmake (hunk 12 above)
Comment (vingu-linaro) [reply product/optee-fvp/fw/Firmware.cmake:22]: ok

**Comment by vingu-linaro** [comment]
Hi All,

I have pushed a new version of the PR which includes changes requested from last comments

File: product/optee-stm32mp1/fw/config_scmi_voltage_domains.c [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+static const struct mod_stm32_pwr_regu_dev_config stm32_pwr_cfg[] = {
+    [STM32_PWR_REG11] = { .pwr_id = PWR_REG11, .regu_name = "reg11" },
+    [STM32_PWR_REG18] = { .pwr_id = PWR_REG18, .regu_name = "reg18" },
+    [STM32_PWR_USB33] = { .pwr_id = PWR_USB33, .regu_name = "usb33" },
+};
+
+/*
+ * stm32_pmic_cfg - Configuration data for PMIC regulators exposed thru SCMI
+ *
+ * These configation data are referenced in the fwk config data of
```
Comment (katvin01) [inline product/optee-stm32mp1/fw/config_scmi_voltage_domains.c:49]: Typo 

I realise "data" is a plural but normally I would opt for "this configuration data is referenced", even though I know it is a plural really :)

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    return FWK_SUCCESS;
+}
+
+static int pmic_regu_get_level(fwk_id_t dev_id, int *level_uv)
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:161]: ditto

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+    return FWK_SUCCESS;
+}
+
+static int pmic_regu_set_level(fwk_id_t dev_id, int level_uv)
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:179]: ditto - and for all below

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+
+    *level_uv = (int32_t)levels[index] * 1000;
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get level PMIC %s = %d",
+         fwk_id_get_element_idx(dev_id), ctx->regu_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_stm32_pmic_regu = {
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:273]: For ease of finding things, maybe order the functions above this in the order they appear within this struct

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    *level_uv = get_regu_voltage(ctx->regu_id);
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:165]: check if level_uv null first, perhaps? (ditto to other uses of it below)

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+{
+    struct stm32_pmic_regu_dev_ctx *ctx;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    *mode_type = MOD_VOLTD_MODE_TYPE_ARCH;
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:112]: check if mod_type and mode_id are null before using them, perhaps?

File: product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+            *max = levels[n];
+        }
+    }
+
+    /* Convert from mV to uV */
+    *min *= 1000;
+    *max *= 1000;
+}
+
+static int pmic_regu_get_info(fwk_id_t dev_id, struct mod_voltd_info *info)
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c:219]: check if info is null, perhaps?

File: product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+ * Voltage domain driver API functions
+ */
+static int pwr_regu_get_config(fwk_id_t dev_id, uint8_t *mode_type,
+                               uint8_t *mode_id)
+{
+    struct stm32_pwr_regu_dev_ctx *ctx;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c:56]: same as above, would do this before using the ID

File: product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+
+static int32_t pwr_regu_level(enum pwr_regulator pwr_id)
+{
+    return (int32_t)stm32mp1_pwr_regulator_mv(pwr_id) * 1000;
+}
+
+/*
+ * Voltage domain driver API functions
+ */
+static int pwr_regu_get_config(fwk_id_t dev_id, uint8_t *mode_type,
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c:49]: check if mode_type and mode_id is null

File: product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+                                     mode_id == MOD_VOLTD_MODE_ID_ON);
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: set_config PWR#%u %s",
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id,
+         mode_id == MOD_VOLTD_MODE_ID_ON ? "on" : "off");
+
+    return FWK_SUCCESS;
+}
+
+static int pwr_regu_get_level(fwk_id_t dev_id, int *level_uv)
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c:106]: check if not null first?  - same for others below too

File: product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+
+    *level_uv = pwr_regu_level(ctx->pwr_id);
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get_level_from_index PWR#%u = %"PRId32,
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_stm32_pwr_regu = {
```
Comment (katvin01) [inline product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c:206]: I would pop the functions in the order they appear here, for ease of finding things in future :)

**Review by katvin01 — CHANGES_REQUESTED** [review]

**Review by tarek-arm — APPROVED** [review]

**Review by tarek-arm — COMMENTED** [review]
Sorry not the PR

**Comment by tarek-arm** [comment]
Jenkins: OK to test

**Comment by vingu-linaro** [comment]
Hi All,

I have just pushed a new version of the PR with @etienne-lms 's changes to take into account @katvin01 's comments on stm32 modules.

File: arch/none/optee/src/arch_main.c [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+
+    optee_mbx_signal_msg_message(device_id, in_buf, in_size, out_buf, out_size);
+
+    fwk_process_event_queue();
+
+    fwk_log_flush();
+#endif
+}
+
+void *aligned_alloc(size_t alignment, size_t size)
```
Comment (etienne-lms) [inline arch/none/optee/src/arch_main.c:95]: Hi Vincent,
OP-TEE OS is being changed to provide this standard function implementation and declaration. Could you remove `aligned_alloc()` from _arch_main.c_?
I can remove also inclusion of _stdlib.h_ and _malloc.h_ at lines 19/20.

**Comment by girishpathak** [comment]
@vingu-linaro Hi Vincent , did you try building on latest master ? as you may have noticed 'smt' is removed from the codebase and replaced by transport module, which is similar but some additional features.  would you see this [module/optee: Create OPTEE shared memory module](https://github.com/ARM-software/SCP-firmware/pull/664/commits/c69fd05168ff0ae8594a419e2ee8b5502999a7aa) need update ?

File: arch/none/optee/src/arch_main.c (hunk 26 above)
Comment (vingu-linaro) [reply arch/none/optee/src/arch_main.c:95]: ok, will do

**Comment by vingu-linaro** [comment]
> @vingu-linaro Hi Vincent , did you try building on latest master ? as you may have noticed 'smt' is removed from the codebase and replaced by transport module, which is similar but some additional features. would you see this [module/optee: Create OPTEE shared memory module](https://github.com/ARM-software/SCP-firmware/pull/664/commits/c69fd05168ff0ae8594a419e2ee8b5502999a7aa) need update ?

Hi @girishpathak , I have noticed this new transport module. Updating Optee module would require to redo a full dev, integration and regression tests cycle so I would prefer to keep this PR as it is. We still have notification support to add for optee arch and we can consider this evolution as part of this next dev cycle.

**Comment by girishpathak** [comment]
Jenkins: ok to test

**Comment by girishpathak** [comment]
jenkins: ok to test

**Comment by vingu-linaro** [comment]
The latest update is only related to the integration of SCP-firmware with optee_os : location in the tree ...

**Comment by girishpathak** [comment]
@vingu-linaro Hi Vincent,
The PR is failing because of code has tabs

```
63 : product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c 
114 : product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c has tab

```
Could you please re-push PR with the above fixed so that it can pass CI?

You can run tools/check_tabs.py to validate.

**Comment by vingu-linaro** [comment]
> @vingu-linaro Hi Vincent, The PR is failing because of code has tabs
> 
> ```
> 63 : product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c 
> 114 : product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c has tab
> ```
> 
> Could you please re-push PR with the above fixed so that it can pass CI?
> 
> You can run tools/check_tabs.py to validate.

Hi @girishpathak ,

Sorry I haven't run tools/check_* on the latest update. I'm going to fix it right now.

**Review by girishpathak — APPROVED** [review]

**Pull request merged**